evm_arithmetization/src/generation/state.rs

```diff
@@ -144,39 +144,73 @@ pub(crate) trait State<F: Field> {
 
     fn is_preinitialized_segment(&self, segment: usize) -> bool;
 
+    fn update_interpreter_final_registers(&mut self, final_registers: RegistersState) {}
+
+    fn get_full_memory(&self) -> Option<MemoryState> {
+        None
+    }
+
     /// Simulates the CPU. It only generates the traces if the `State` is a
     /// `GenerationState`.
-    fn run_cpu(&mut self) -> anyhow::Result<()>
+    fn run_cpu(
+        &mut self,
+        max_cpu_len_log: Option<usize>,
+    ) -> anyhow::Result<(RegistersState, Option<MemoryState>)>
     where
         Self: Transition<F>,
-        Self: Sized,
     {
+        println!("max cpu len log {:?}", max_cpu_len_log);
         let halt_offsets = self.get_halt_offsets();
 
+        let halt_pc = KERNEL.global_labels["halt"];
+        let halt_final_pc = KERNEL.global_labels["halt_final"];
+        let mut final_registers = RegistersState::default();
+        let mut final_mem = self.get_mut_generation_state().memory.clone();
+        let mut running = true;
+
         loop {
             let registers = self.get_registers();
             let pc = registers.program_counter;
 
-            let halt = registers.is_kernel && halt_offsets.contains(&pc);
+            let halt_final = registers.is_kernel && halt_offsets.contains(&pc);
+            if running
+                && (registers.is_kernel && pc == halt_pc
+                    || (max_cpu_len_log.is_some()
+                        && self.get_clock()
+                            == (1 << max_cpu_len_log.unwrap()) - NUM_EXTRA_CYCLES_AFTER))
```

Comment:
> Also, `max_cpu_len_log == None` should just make the condition false

Sorry, I'm not sure I understand what you mean with this...

Reply:
Because there's an unwrap, but i just realized it will alsways work or the condition would be false. However, it might still be a good idea to get rid of the unwrap (and wrap the conditions with some methods)